src/sinks/aws_cloudwatch_logs/request.rs

```diff
@@ -0,0 +1,159 @@
+    client: Client,
+    state: State,
+    events: Option<Vec<InputLogEvent>>,
+    token_tx: Option<oneshot::Sender<Option<String>>>,
+}
+
+struct Client {
+    client: CloudWatchLogsClient,
+    stream_name: String,
+    group_name: String,
+}
+
+enum State {
+    CreateStream(RusotoFuture<(), CreateLogStreamError>),
+    DescribeStream(RusotoFuture<DescribeLogStreamsResponse, DescribeLogStreamsError>),
+    Put(RusotoFuture<PutLogEventsResponse, PutLogEventsError>),
+}
+
+impl CloudwatchFuture {
+    pub fn new(
+        client: CloudWatchLogsClient,
+        stream_name: String,
+        group_name: String,
+        events: Vec<InputLogEvent>,
+        token: Option<String>,
+        token_tx: oneshot::Sender<Option<String>>,
+    ) -> Self {
+        let client = Client {
+            client,
+            stream_name,
+            group_name,
+        };
+
+        let (state, events) = if let Some(token) = token {
+            let state = State::Put(client.put_logs(Some(token), events));
+            (state, None)
+        } else {
+            let state = State::DescribeStream(client.describe_stream());
+            (state, Some(events))
+        };
+
+        Self {
+            client,
+            events,
+            state,
+            token_tx: Some(token_tx),
+        }
+    }
+}
+
+impl Future for CloudwatchFuture {
+    type Item = ();
+    type Error = CloudwatchError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match &mut self.state {
+                State::DescribeStream(fut) => {
+                    let response = try_ready!(fut.poll().map_err(CloudwatchError::Describe));
+
+                    if let Some(stream) = response
+                        .log_streams
+                        .ok_or(CloudwatchError::NoStreamsFound)?
+                        .into_iter()
+                        .next()
+                    {
+                        trace!(message = "stream found", stream = ?stream.log_stream_name);
+
+                        let events = self
+                            .events
+                            .take()
+                            .expect("Token got called twice, this is a bug!");
+
+                        let token = stream.upload_sequence_token;
+
+                        trace!(message = "putting logs.", ?token);
+                        self.state = State::Put(self.client.put_logs(token, events));
+                    } else {
+                        trace!("provided stream does not exist; creating a new one.");
+                        self.state = State::CreateStream(self.client.create_log_stream());
+                    }
+                }
+
+                State::CreateStream(fut) => {
+                    let _ = try_ready!(fut.poll().map_err(CloudwatchError::CreateStream));
+
+                    trace!("stream created.");
+
+                    self.state = State::DescribeStream(self.client.describe_stream());
```

Comment:
Don't we want to go straight to `Put` here?

Reply:
Currently, there are no clear documentation about how a token is treated at the start of a stream. For this case I think its safe to always call describestreams to ensure we always call put with a valid token.